### PR TITLE
Clear caches from more places on a hot reload event

### PR DESCRIPTION
**Pull request by pranavkm**

### CODE REVIEW

**Comment by pranavkm** [comment]
FYI @tommcdon \ @mikem8361

**Review by javiercn — APPROVED** [review]
LGTM.

Is there a way we could centralize the "clear" behavior into a cache type?

**Comment by pranavkm** [comment]
> Is there a way we could centralize the "clear" behavior into a cache type?

I like this. It's less likely we would add a new cache in the future and forget to clear it up if we do this.

File: src/Components/Components/src/Routing/Router.cs [hunk 1]
```diff
@@ -152,6 +162,12 @@ private void RefreshRouteTable()
             }
         }
 
+        private void ClearRouteCaches()
+        {
+            RouteTableFactory.ClearCaches();
+            _currentRouteKey = default;
```
[the remark this answers is not in the document]
Comment (SteveSandersonMS) [reply src/Components/Components/src/Routing/Router.cs]: In fact - I hope you don't mind - I've added a commit to this PR with that change. Hopefully CI will have gone through by the time you see this and it won't slow you down at all.

File: src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs [hunk 2]
```diff
@@ -426,6 +429,12 @@ private static Assembly GetRequiredLoadedAssembly(AssemblyKey assemblyKey)
                 ?? throw new ArgumentException($"There is no loaded assembly with the name '{assemblyKey.AssemblyName}'.");
         }
 
+        private static void ClearCache(Type[]? _)
+        {
+            _cachedMethodsByAssembly.Clear();
+            _cachedMethodsByType.Clear();
+        }
```
Comment (SteveSandersonMS) [inline src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs:436]: Did you forget to call this method from somewhere, or am I just missing where it's called from?

**Review by SteveSandersonMS — APPROVED** [review]

**Comment by SteveSandersonMS** [comment]
@eerhardt I just added an approval here and didn't anticipate that was going to trigger an auto-merge even though there was still one of your feedback items pending: https://github.com/dotnet/aspnetcore/pull/32506/files#r629441507

Do you still want a follow-up on that one?

Auto-merge is a bit of an awkward thing if the PR isn't already approved, because it means subsequent reviewers have to (1) check if `auto-merge` is present and if so, (2) make sure there's no pending feedback from anyone else before marking the PR as approved.

**Comment by pranavkm** [comment]
While centralizing the caches sounds nice on paper, it's a bit difficult. The caches are on a mix of instance and static types which makes the cache lifetime hard to pin down. Additionally, they're also split across several assemblies that do not have IVT access to each other. We definitely don't want to introduce public APIs for caches. I'm going to punt on doing that for now. MVC might be a bit more open to this option since all of it's caches are non-static.

**Pull request merged**